### PR TITLE
CHAP-4 Factor view into master page and create class User class

### PR DESCRIPTION
**Pull request by LostSheep**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,3 +29,6 @@
 
 # Ignore master key for decrypting credentials and more.
 /config/master.key
+
+# Ignore Spring files.
+/spring/*.pid
```
Comment (bwang-ith) [inline .gitignore]: You've got a bad character at the end of the file See how GH points it out for you. This will be automatically solved by using vim. I know it slows you down in the short term, but it speeds you up in the long term.

File: Gemfile [hunk 2]
```diff
@@ -22,6 +23,14 @@ group :development do
   gem 'spring-watcher-listen'
 end
 
+group :test do
+  gem 'rails-controller-testing'
+  gem 'minitest'
+  gem 'minitest-reporters'
+  gem 'guard'
```
Comment (bwang-ith) [inline Gemfile]: I thought you decided not to use guard. As a rule of thumb, minimize the the number of external dependencies. Never add a gem for reasons like "why not?"

File: Guardfile [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# A sample Guardfile
```
Comment (bwang-ith) [inline Guardfile]: Don't need guard file without guard.

File: User_class.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# freeze string literal: true
+
+# User view model.
+class User
+  attr_accessor :firstName, :lastName, :email
```
Comment (bwang-ith) [inline User_class.rb]: you don't need these. Also, we use snake_case, not camelCase for vars.

File: app/controllers/microposts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+class MicropostsController < ApplicationController
+  before_action :set_micropost, only: %i[show edit update destroy]
+
+  # GET /microposts.
+  # List microposts.
+  def index
+    @microposts = Micropost.all
+  end
+
+  # GET /microposts/new
```
Comment (bwang-ith) [inline app/controllers/microposts_controller.rb]: period.

Comment (bwang-ith) [reply app/controllers/microposts_controller.rb]: same for others.

File: app/controllers/static_pages_controller.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Handles static page for home and help.
+class StaticPagesController < ApplicationController
+  
+  # Get home.
+  def home;  end
```
Comment (bwang-ith) [inline app/controllers/static_pages_controller.rb]: I don't think you need any of these methods. This can be an empty class. If I'm wrong, then at least you need to comment them properly.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+# Handle user CRUD.
+class UsersController < ApplicationController
+  before_action :set_user, only: %i[show edit update destroy]
+
+  # GET /users.
+  # List Users.
+  def index
+    @users = User.all
+  end
+
```
Comment (bwang-ith) [inline app/controllers/users_controller.rb:12]: Remove extra blank line

File: app/helpers/static_pages_helper.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# freeze_string_literals: true
+
+# Help static_pages view
+module StaticPagesHelper
+
```
Comment (bwang-ith) [inline app/helpers/static_pages_helper.rb]: Remove blank line

File: app/helpers/static_pages_helper.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# freeze_string_literals: true
+
+# Help static_pages view
+module StaticPagesHelper
+  
+  # Return full http title
```
Comment (bwang-ith) [inline app/helpers/static_pages_helper.rb]: period

File: app/helpers/users_helper.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module UsersHelper
```
Comment (bwang-ith) [inline app/helpers/users_helper.rb:4]: comment the module. All classes, modules, methods get comments.

File: app/models/micropost.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# Validates Micropost
+class Micropost < ApplicationRecord
+  belongs_to :user
+  validates :content, length: { maximum: 140 },
```
Comment (bwang-ith) [inline app/models/micropost.rb]: Create a file called lib/constants.rb and extract 140 to a constant. This is how we structure all our apps. See IO and DM for example and do it the same.

File: app/views/microposts/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,29 @@
+<p id="notice"><%= notice %></p>
```
Comment (bwang-ith) [inline app/views/microposts/index.html.erb]: single quotes. same for whole page.

File: app/views/static_pages/about.html.erb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, "About") %>
```
Comment (bwang-ith) [inline app/views/static_pages/about.html.erb]: same

File: app/views/static_pages/home.html.erb [hunk 14]
```diff
@@ -0,0 +1,19 @@
+<% provide(:title, "Home") %>
```
Comment (bwang-ith) [inline app/views/static_pages/home.html.erb]: same

File: app/views/users/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,29 @@
+<p id="notice"><%= notice %></p>
```
Comment (bwang-ith) [inline app/views/users/index.html.erb]: you get it

File: config/routes.rb [hunk 16]
```diff
@@ -1,3 +1,11 @@
+# frozen_string_literal: true
+
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :microposts
+  resources :users
+  
+  root 'static_pages#home'
+  get 'static_pages/home'
```
Comment (bwang-ith) [inline config/routes.rb:8]: This works? I've never seen a route declaration without a `to:` argument. Ah, the magic of Rails...

File: test/controllers/microposts_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+require 'test_helper'
+
+class MicropostsControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @micropost = microposts(:one)
+    @user = users(:one)
+  end
+  
+  # Test index
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: period

File: test/controllers/microposts_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+  test 'should succeed get index' do
+    get microposts_path
+    assert_response :success
+  end
+
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb:21]: Add `# Test new.`

File: test/controllers/microposts_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+  test 'should succeed get new' do
+    get new_micropost_path
+    assert_response :success
+  end
+
+# Test create
+
+  test 'should succeed post create microtest' do
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: What's a microtest?

File: test/controllers/microposts_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    # assert_difference('Micropost.count', 1) do
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: Uncomment code

File: test/controllers/microposts_controller_test.rb [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+      post microposts_path, params: { micropost: { content: @micropost.content, user_id: @user.id } }
+    # end
+    assert_redirected_to micropost_path(Micropost.last)
+  end
+
+  # Test Update
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: Period. same for rest.

File: test/controllers/microposts_controller_test.rb [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+  test 'should succeed update micropost' do
+
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: Remove blank line

File: test/controllers/microposts_controller_test.rb [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    # end
+    assert_redirected_to micropost_path(Micropost.last)
+  end
+
+  # Test Update
+
+  test 'should succeed update micropost' do
+    
+    patch micropost_path(@micropost), params: 
+    {
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: indent this hash. I think rubocop will have suggestions for this.

File: test/controllers/microposts_controller_test.rb [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+      micropost: 
+      { 
+        content: @micropost.content, 
+        user_id: @micropost.user_id
+      }, format: :html
+    }
+    assert_redirected_to micropost_path(@micropost)
+  end
+
+  # Test Delete
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: `Delete` => `destroy.` (we're testing the _action_, not the HTTP verb.

File: test/controllers/microposts_controller_test.rb [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+  test 'should succeed destroy micropost' do
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: `should succeed delete destroy`

File: test/controllers/static_pages_controller_test.rb [hunk 26]
```diff
@@ -0,0 +1,34 @@
+require 'test_helper'
+
+class StaticPagesControllerTest < ActionDispatch::IntegrationTest
+
```
Comment (bwang-ith) [inline test/controllers/static_pages_controller_test.rb]: remove blank line

File: test/controllers/static_pages_controller_test.rb [hunk 27]
```diff
@@ -0,0 +1,34 @@
+require 'test_helper'
+
+class StaticPagesControllerTest < ActionDispatch::IntegrationTest
```
Comment (bwang-ith) [inline test/controllers/static_pages_controller_test.rb:6]: comment class

File: test/controllers/static_pages_controller_test.rb [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+  
+  # Test setup
+  def setup
+    @base_title = 'Ruby on Rails Tutorial Sample App'
+  end
+  
+  # Test index
```
Comment (bwang-ith) [inline test/controllers/static_pages_controller_test.rb]: Delete dead code. (or wake it from the dead!)

File: test/controllers/static_pages_controller_test.rb [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+  test 'should succeed get help' do
+    get static_pages_help_path
+    assert_response :success
+    assert_select 'title', "Help | #{@base_title}"
+  end
+  
+  test 'should succeed get about' do
+    get static_pages_about_path
+    assert_response :success
+    assert_select 'title', "About | #{@base_title}"
```
Comment (bwang-ith) [inline test/controllers/static_pages_controller_test.rb:40]: Above everything, be consistent. So if you're going to `assert_select` the title here, do it in the Micropost controller tests for index, show, edit

File: test/controllers/users_controller_test.rb [hunk 30]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
```
Comment (bwang-ith) [inline test/controllers/users_controller_test.rb:6]: comment, same for all classes.

File: test/controllers/users_controller_test.rb [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+  setup do
+    @user = users(:one)
+  end
+
+  test 'should get index' do
```
Comment (bwang-ith) [inline test/controllers/users_controller_test.rb]: succeed/fail, same for others

File: test/fixtures/microposts.yml [hunk 32]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  content: MyText
```
Comment (bwang-ith) [inline test/fixtures/microposts.yml]: More realistic attribute values are often useful in fixtures.

File: test/fixtures/users.yml [hunk 33]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  name: MyString
```
Comment (bwang-ith) [inline test/fixtures/users.yml]: same

File: test/models/micropost_test.rb [hunk 34]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+# Micropost model test class. 
+class MicropostTest < ActiveSupport::TestCase
+
```
Comment (bwang-ith) [inline test/models/micropost_test.rb]: remove blank line

File: test/models/micropost_test.rb [hunk 35]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+# Micropost model test class.
```
Comment (bwang-ith) [inline test/models/micropost_test.rb]: not a sentence

File: test/models/micropost_test.rb [hunk 36]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+# Micropost model test class. 
+class MicropostTest < ActiveSupport::TestCase
+    
+    # Setup for tests.
+    def setup
+       @user = users(:one)
```
Comment (bwang-ith) [inline test/models/micropost_test.rb]: fix indents

File: test/models/micropost_test.rb [hunk 37]
```diff
@@ -0,0 +1,42 @@
...
+    end
+    
+    test 'should succeed empty micropost content' do
+        micropost = Micropost.new(content: '', user: @user)
```
Comment (bwang-ith) [inline test/models/micropost_test.rb]: fix indents. same for whole file

File: test/models/user_test.rb [hunk 38]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+    test 'should succeed empty user name' do
```
Comment (bwang-ith) [inline test/models/user_test.rb]: fix indents, same for whole file

File: test/system/users_test.rb [hunk 39]
```diff
@@ -0,0 +1,47 @@
...
+
+require 'application_system_test_case'
+
+class UsersTest < ApplicationSystemTestCase
+  setup do
+    @user = users(:one)
+  end
+
+  test 'visiting the index' do
+    visit users_url
```
Comment (bwang-ith) [inline test/system/users_test.rb]: hm... never seen `visit` before, did Rails give you that?

**Review by bwang-ith — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 2 above)
Comment (LostSheep) [reply Gemfile]: We talked about it but I forgot that it was already added. Will remove.

File: .gitignore (hunk 1 above)
Comment (LostSheep) [reply .gitignore]: Bad character?
I figured id learn ruby first, and then get use to vi but i can start using it.

File: User_class.rb (hunk 4 above)
Comment (LostSheep) [reply User_class.rb]: Old habit, will change.

File: test/system/users_test.rb (hunk 39 above)
Comment (LostSheep) [reply test/system/users_test.rb]: I haven't touched the system test/system folder so must be auto generated.

File: Gemfile [hunk 40]
```diff
@@ -1,33 +1,37 @@
...
+  gem 'minitest-reporters'
 end
 
 group :production do
   gem 'pg'
 end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
-gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
+gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
```
Comment (bwang-ith) [inline Gemfile]: Bad newline char. Make sure to check your diff on GitHub so that you can easily see these.

File: Gemfile [hunk 41]
```diff
@@ -1,33 +1,37 @@
...
-gem 'jbuilder'
-gem 'jquery-rails'
-gem 'puma'
 gem 'rails'
-gem 'rubocop'
+gem 'puma'
 gem 'sass-rails'
-gem 'turbolinks'
 gem 'uglifier'
+gem 'coffee-rails'
```
Comment (bwang-ith) [inline Gemfile]: order your gems in alphabetical order.

File: Gemfile.lock [hunk 42]
```diff
@@ -61,8 +63,22 @@ GEM
     erubi (1.7.1)
     execjs (2.7.0)
     ffi (1.9.25)
+    formatador (0.2.5)
     globalid (0.4.1)
       activesupport (>= 4.2.0)
+    guard (2.15.0)
```
Comment (bwang-ith) [inline Gemfile.lock]: Did you `bundle` your gems?

File: User_class.rb [hunk 43]
```diff
@@ -0,0 +1,28 @@
+# freeze string literal: true
+
+# User view model.
+class User
+  attr_accessor :first_Name, :last_Name, :email
```
Comment (bwang-ith) [inline User_class.rb]: `:first_Name` => `:first_name`, same for last

File: User_class.rb [hunk 44]
```diff
@@ -0,0 +1,28 @@
...
+  
+  # Init user
```
Comment (bwang-ith) [inline User_class.rb]: period

File: User_class.rb (hunk 43 above)
Comment (bwang-ith) [reply User_class.rb]: Also, I'm a bit confused about the reason behind adding these `*_name` attributes. Since they are not defined in the db schema, they will not persist past the length of a request. I would suggest either not using them at all, or updating your db migration to spit up `name` into `first_name` and `last_name`.

Also, you definitely don't need `email` there since it _is_ in the schema

File: app/controllers/microposts_controller.rb [hunk 45]
```diff
@@ -22,42 +22,42 @@ def create
...
   end
 
-  # PATCH/PUT /microposts/1
+  # PATCH/PUT /microposts/1.
   # Update micropost.
   def update
     respond_to do |format|
       if @micropost.update(micropost_params)
-        format.html do
+        format.html {
```
Comment (bwang-ith) [inline app/controllers/microposts_controller.rb]: revert these changes. Multi-line blocks get `do`/`end` syntax, curly braces are reserved for single-line blocks. That's from rubocop.

**Review by bwang-ith — CHANGES_REQUESTED** [review]

File: Gemfile.lock (hunk 42 above)
Comment (LostSheep) [reply Gemfile.lock]: It didn't occur to me to bundle the gems before pushing. Should this be in a .gitignore?

File: User_class.rb (hunk 43 above)
Comment (LostSheep) [reply User_class.rb]: The tutorial said to create it this way then went on to say it would be used in the next chapter. I don't know the reasoning for it but I'm trusting the tutorial has a plan. Likewise I assume classes are typically added to the root directory of an application.

File: Guardfile (hunk 3 above)
Comment (LostSheep) [reply Guardfile]: fixed

File: app/controllers/static_pages_controller.rb (hunk 6 above)
Comment (LostSheep) [reply app/controllers/static_pages_controller.rb]: is "# Get static_pages/home." the correct commenting?

File: app/controllers/users_controller.rb (hunk 7 above)
Comment (LostSheep) [reply app/controllers/users_controller.rb:12]: When edited with vi there is no extra line.

File: app/controllers/microposts_controller.rb [hunk 46]
```diff
@@ -22,23 +22,23 @@ def create
...
     respond_to do |format|
       if @micropost.save
-        format.html do
+        format.html {
           redirect_to @micropost,
                       notice: 'Micropost was successfully created.'
-        end
-        format.json do
+        }
+        format.json {
```
Comment (bwang-ith) [inline app/controllers/microposts_controller.rb]: revert all these changes. Curly braces are reserved for one-liner blocks

File: app/controllers/static_pages_controller.rb (hunk 6 above)
Comment (bwang-ith) [reply app/controllers/static_pages_controller.rb]: Get -> GET
just see any of the controllers in IO/DM

File: app/controllers/static_pages_controller.rb [hunk 47]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Handles static page for home and help.
+class StaticPagesController < ApplicationController
+
```
Comment (bwang-ith) [inline app/controllers/static_pages_controller.rb]: remove blank line

File: lib/constants.rb [hunk 48]
```diff
@@ -0,0 +1,5 @@
+# Module for constants.
+module Constants
+  # Maximum length of a micropost
+  MICROPOST_MAX_STRING_LEN = 140
+end
```
Comment (bwang-ith) [inline lib/constants.rb]: bad char

File: test/controllers/microposts_controller_test.rb [hunk 49]
```diff
@@ -2,51 +2,57 @@
 
 require 'test_helper'
 
+# Test Micropost controler CRUD actions.
 class MicropostsControllerTest < ActionDispatch::IntegrationTest
   setup do
     @micropost = microposts(:one)
     @user = users(:one)
+    @base_title = 'Ruby on Rails Tutorial Sample App'
```
Comment (bwang-ith) [inline test/controllers/microposts_controller_test.rb]: Extract string to lib/constants.rb

**Review by bwang-ith — CHANGES_REQUESTED** [review]

**Review by bwang-ith — APPROVED** [review]
